allennlp/models/encoder_decoders/copynet_seq2seq.py

```diff
@@ -0,0 +1,824 @@
+import logging
+from typing import Dict, Tuple, List, Any, Union
+
+import numpy
+from overrides import overrides
+import torch
+from torch.nn.modules.linear import Linear
+from torch.nn.modules.rnn import LSTMCell
+
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.util import START_SYMBOL, END_SYMBOL
+from allennlp.data.vocabulary import Vocabulary
+from allennlp.models.model import Model
+from allennlp.modules import Attention, TextFieldEmbedder, Seq2SeqEncoder
+from allennlp.modules.token_embedders import Embedding
+from allennlp.nn import util
+from allennlp.training.metrics import Metric, BLEU
+from allennlp.nn.beam_search import BeamSearch
+
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@Model.register("copynet_seq2seq")
+class CopyNetSeq2Seq(Model):
+    """
+    This is an implementation of `CopyNet <https://arxiv.org/pdf/1603.06393>`_.
+    CopyNet is a sequence-to-sequence encoder-decoder model with a copying mechanism
+    that can copy tokens from the source sentence into the target sentence instead of
+    generating all target tokens only from the target vocabulary.
+
+    It is very similar to a typical seq2seq model used in neural machine translation
+    tasks, for example, except that in addition to providing a "generation" score at each timestep
+    for the tokens in the target vocabulary, it also provides a "copy" score for each
+    token that appears in the source sentence. In other words, you can think of CopyNet
+    as a seq2seq model with a dynamic target vocabulary that changes based on the tokens
+    in the source sentence, allowing it to predict tokens that are out-of-vocabulary (OOV)
+    with respect to the actual target vocab.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``, required
+        Vocabulary containing source and target vocabularies.
+    source_embedder : ``TextFieldEmbedder``, required
+        Embedder for source side sequences
+    encoder : ``Seq2SeqEncoder``, required
+        The encoder of the "encoder/decoder" model
+    attention : ``Attention``, required
+        This is used to get a dynamic summary of encoder outputs at each timestep
+        when producing the "generation" scores for the target vocab.
+    beam_size : ``int``, required
+        Beam width to use for beam search prediction.
+    max_decoding_steps : ``int``, required
+        Maximum sequence length of target predictions.
+    target_embedding_dim : ``int``, optional (default = 30)
+        The size of the embeddings for the target vocabulary.
+    copy_token : ``str``, optional (default = '@COPY@')
+        The token used to indicate that a target token was copied from the source.
+        If this token is not already in your target vocabulary, it will be added.
```

Comment:
Looks like you crash instead of adding it.  Which behavior is correct?  Seems nicer to just add it instead of crashing.  Did that not work?

Reply:
If I recall correctly, the first way I had this implemented I was actually adding the copy token into the vocab's source namespace when the model was instantiated. But doing so had some unintended consequences. I can't quite remember the issue it caused, but I think it was because the serialized vocabulary (in the model serialization directory) would not include the copy token in the source namespace.

Reply:
Ah, yes, that would have been a problem.  It has since been fixed (about 3 months ago; see [here](https://github.com/allenai/allennlp/pull/1857/files#diff-d4e8f335bf383d8d2b18287949003bbeR276)), so it _should_ be ok to just add it now.

Reply:
Ah thanks! Yeup, that should do it.